evm_arithmetization/src/keccak/round_flags.rs

```diff
@@ -56,19 +60,25 @@ pub(crate) fn eval_round_flags_recursively<F: RichField + Extendable<D>, const D
     }
 
     // Flags should circularly increment, or be all zero for padding rows.
+    let current_any_flag =
+        builder.add_many_extension((0..NUM_ROUNDS).map(|i| local_values[reg_step(i)]));
     let next_any_flag =
         builder.add_many_extension((0..NUM_ROUNDS).map(|i| next_values[reg_step(i)]));
+    let last_row_flag = local_values[reg_step(NUM_ROUNDS - 1)];
     for i in 0..NUM_ROUNDS {
         let current_round_flag = local_values[reg_step(i)];
         let next_round_flag = next_values[reg_step((i + 1) % NUM_ROUNDS)];
-        let diff = builder.sub_extension(next_round_flag, current_round_flag);
-        let constraint = builder.mul_extension(next_any_flag, diff);
+        let diff1 = builder.sub_extension(next_round_flag, current_round_flag);
+        let constraint1 = builder.mul_extension(next_any_flag, diff1);
+        let diff2 = builder.sub_extension(next_any_flag, one);
+        let diff3 = builder.sub_extension(last_row_flag, one);
+        let prod = builder.mul_extension(diff2, diff3);
+        let constraint2 = builder.mul_extension(current_any_flag, prod);
+        let constraint = builder.add_extension(constraint1, constraint2);
```

Comment:
nit: we can reduce the number of constraints added to the circuits here, leveraging that some multiplicands are of the form `foo - 1`, hence calling `builder.mul_sub_extension`.
I think we could replicate:
```rust
        yield_constr.constraint_transition(
            next_any_flag * (next_round_flag - current_round_flag)
                + (next_any_flag - F::ONE) * current_any_flag * (last_round_flag - F::ONE),
        );
```
here with

```suggestion
        let flag_diff = builder.sub_extension(next_round_flag, current_round_flag);
        let constraint1 = builder.mul_extension(next_any_flag, flag_diff);
        let constraint2 = {
            let tmp = builder.mul_sub_extension(current_any_flag, next_any_flag, current_any_flag);
            builder.mul_sub_extension(tmp, last_round_flag, tmp)
        };
        let constraint = builder.add_extension(constraint1, constraint2);
```
(5x25 constraints instead of 7x25)
Note that I have renamed one register to `last_round_flag` as per suggestion above.

Reply:
Actually, in that case, I think we can move the additional constraint out of the loop itself
```rust
    // Flags should circularly increment, or be all zero for padding rows.
    let current_any_flag =
        builder.add_many_extension((0..NUM_ROUNDS).map(|i| local_values[reg_step(i)]));
    let next_any_flag =
        builder.add_many_extension((0..NUM_ROUNDS).map(|i| next_values[reg_step(i)]));
    let last_round_flag = local_values[reg_step(NUM_ROUNDS - 1)];
    let base_constraint = {
        let tmp = builder.mul_sub_extension(current_any_flag, next_any_flag, current_any_flag);
        builder.mul_sub_extension(tmp, last_round_flag, tmp)
    };
    for i in 0..NUM_ROUNDS {
        let current_round_flag = local_values[reg_step(i)];
        let next_round_flag = next_values[reg_step((i + 1) % NUM_ROUNDS)];
        let flag_diff = builder.sub_extension(next_round_flag, current_round_flag);
        let constraint = builder.mul_add_extension(next_any_flag, flag_diff, base_constraint);
        yield_constr.constraint_transition(builder, constraint);
    }
```
Not sure about the variable names but wdyt?

Reply:
Yeah that's indeed better, thanks!